algebird-core/src/main/scala/com/twitter/algebird/Scan.scala

```diff
@@ -0,0 +1,361 @@
+package com.twitter.algebird
+
+import scala.collection.compat._
+import scala.collection.generic.CanBuildFrom
+
+object Scan {
+  /**
+    * Most consumers of Scan don't care about the type of the {{State}} type variable. But for those that do,
+    * we make an effort to expose it in all of our combinators.
+    * @tparam I
+    * @tparam S
+    * @tparam O
+    */
+  type Aux[-I, S, +O] = Scan[I, O] { type State = S }
+
+  implicit def applicative[I]: Applicative[({ type L[O] = Scan[I, O] })#L] = new ScanApplicative[I]
+
+  def from[I, S, O](initState: S)(presentAndNextStateFn: (I, S) => (O, S)): Aux[I, S, O] =
+    new Scan[I, O] {
+      override type State = S
+      override val initialState = initState
+      override def presentAndNextState(i: I, s: State): (O, State) = presentAndNextStateFn(i, s)
+    }
+
+  def fromFunction[I, O](f: I => O): Aux[I, Unit, O] = new Scan[I, O] {
+    override type State = Unit
+    override val initialState = ()
+    override def presentAndNextState(i: I, stateBeforeProcessingI: Unit): (O, State) = (f(i), ())
+  }
+
+  /**
+    * Streams can be thought of as being a hidden state that is queryable for a head element, and another hidden state
+    * that represents the rest of the stream. Scans take streams of inputs to streams of outputs, but some scans
+    * have trivial inputs and just produce a stream of outputs.
+    * @param initState The initial state of the scan; think of this as an infinite stream.
+    * @param destructor This function decomposes a stream into the its head-element and tail-stream.
+    * @tparam S The hidden state of the stream that we are turning into a Scan.
+    * @tparam O The type of the elments of the stream that we are turning into a Scan
+    * @return A Scan whose inputs are irrelevant, and whose outputs are those that we would get from implementing
+    *         a stream using the information provided to this method.
+    */
+  def fromStreamLike[S, O](initState: S)(destructor: S => (O, S)): Aux[Any, S, O] = new Scan[Any, O] {
```

Comment:
can we call this `def iterate`? This is very similar to `Iterator.iterate` or `Stream.iterate` in the scala standard library.

This is really awesome.